nixos/modules/services/networking/pangolin/default.nix

```
@@ -0,0 +1,1270 @@
+          ];
+
+        systemd.services = mkClientsAttr (
+          pkg:
+          mkIf cfg.${pkg}.enable {
+            description = packageDescription pkg;
+            wantedBy = [ "multi-user.target" ];
+            after = [ "network.target" ];
+
+            serviceConfig = (commonOptions pkg) // {
+              ExecStart = "${getExe cfg.${pkg}.package} ${toGNUCommandLineShell { } cfg.${pkg}.settings}";
+            };
+          }
+        );
+      })
+      (mkIf (pangolin.enable || gerbil.enable) {
+        # These are needed regardless if Pangolin or Gerbil are enabled,
+        # as they share config options.
+        assertions =
+          # Assertions for ports.
+          (map (
+            type:
+            let
+              illegalPorts =
+                if (type == "tcp") then
+                  httpPorts ++ optionals pangolin.enable internalPorts
+                else if gerbil.enable then
+                  gerbilPorts
+                else
+                  [ ];
+            in
+            {
+              assertion = mutuallyExclusive illegalPorts pangolin.extraPorts.${type};
+              message = ''
+                'services.pangolin.extraPorts.${type}' contains the following illegal ports:
+                - ${concatMapStringsSep "\n- " toString (intersectLists illegalPorts pangolin.extraPorts.${type})}
+              '';
+            }
+          ) (attrNames pangolin.extraPorts))
+          ++ [
+            {
+              assertion = allUnique internalPorts;
+              message = ''
+                The following ports may have conflicting values:
+                - ${
+                  concatMapStringsSep "\n- "
+                    (port: "services.pangolin.settings.server.${port}: ${toString pangolin.settings.server.${port}}")
+                    (
+                      pangolin.enable [
+                        "internal_port"
+                        "external_port"
+                        "next_port"
+                        "integration_port"
+                      ]
+                    )
+                }
+                ${optionalString gerbil.enable "- services.gerbil.port: ${toString gerbil.port}"}
+              '';
+            }
+          ]
+
+          # Assertions for the domain configuration.
+          ++ (
+            let
+              certResolvers = unique ((zipAttrs (attrValues domains)).cert_resolver);
+              domainsMappedToResolvers = map (
+                resolver:
+                attrNames (
+                  filterAttrsRecursive (_: domain: isAttrs domain && domain.cert_resolver == resolver) domains
+                )
+              ) certResolvers;
+              allIdentical = list: all (x: (head list) == x) list;
+              check =
+                attr:
+                map (list: allIdentical list) (
+                  map (listOfDomain: map (domain: domains.${domain}.${attr}) listOfDomain) domainsMappedToResolvers
+                );
```

you didnt think you could sneak this past me did you.... ;)
```suggestion
              check =
                attr:
                  map (allIdentical <| (listOfDomain: map (domain: domains.${domain}.${attr}) listOfDomain)) domainsMappedToResolvers
                );
```
this can for sure be made smaller still, give me a minute

---

Pipe operators **cannot** go into Nixpkgs.

---

I've removed the `list: x list`, but I need a lib.pipe example because I have no idea how the operator works.